### DISCUSSION TITLE
/usr/local/Frameworks vs /Library/Frameworks

### DISCUSSION THREAD

**Discussion by memeplex** [Homebrew/discussions#138] [Everyday usage]
I've installed python using brew and I see the framework is symlinked from the Cellar to /usr/local/Frameworks instead of /Library/Frameworks.

This confuses me because all documentation regarding frameworks I've found suggest to use one of the four Library folders (user, local, system, network) and, in this case, /Library/Frameworks (that is the local Library) seems the most appropriate one and, indeed, is what the official python installer for MacOS uses.

I understand the usage of the standard unix directory structure for *nix packages, but why also for frameworks whose location is well defined and documented by Apple?

**Comment by SMillerDev** [comment]
Because we're not allowed to write there without using sudo

**Comment by memeplex** [comment]
I kind of expected that answer, but casks write to /Applications, don't them? Is it different from writing to /Library? Sorry I'm new to MacOS and, coming from Linux, I'm still trying to make sense of the somewhat split personality of the file system.

**Reply by SMillerDev** [reply]
Casks are special, they just run the installer that upstream provides. If that writes to such directories, so does the cask. Normal formulae only ever install to a subdirectory of `/usr/local`

**Comment by memeplex** [comment]
Ok, I see, makes sense. Then wouldn't it be better to install python as a cask from their official package? There are lots of posts around the web explaining how to completely remove python (Applications, Frameworks, and the symlinks with brew cleanup). That could be automated using cask (OTOH they should provide a proper uninstaller...).

**Reply by carlocab** [reply]
Why would it be better? 

Other installers (presumably also the one from python.org) install bits and pieces of Python all over your system -- that's why there are all these explanations of how to uninstall it. The Homebrew Python formula is completely contained in `/usr/local/Cellar` (on Intel Macs, at least), which makes uninstalling it very easy.

**Comment by memeplex** [comment]
Maybe because it's the official installer but fundamentally because it tries harder to comply with both worlds:
* The framework part goes into /Library (which brew won't do because of its policy)
* The *nix part goes into /usr/local

I find this more coherent than putting the framework into /usr/local, because that's the way it has been sanctioned.

OTOH the official installer fails to provide a sane uninstall procedure, and here brew would shine.

**Reply by carlocab** [reply]
The location of the Frameworks directory is an implementation detail that shouldn't matter for a user who doesn't mess around with Python internals.

The python.org installer uses `/Library/Frameworks` because that's where the built-in macOS Python used to be. It is now in `/Library/Python`. Apple could change it because the location doesn't really matter, as long as your Python is configured so that it knows where to find that directory.

So, AFAICT, there is absolutely no advantage to using `/Library/Frameworks`, and potentially only disadvantages (messy uninstallation). If using an alternative location somehow causes Python to misbehave, that's because of a poorly configured installation.

**Comment by memeplex** [comment]
Ok, thanks for the clarification, as I said before I'm very new to the platform and I had assumed there was some tooling, perhaps from Apple, that by default looked for frameworks in the documented location, so installing them to /usr/local would have required further tweaks from the user.

**Reply by carlocab** [reply]
If there some post-install tweaks that you need to do when `brew` installs a package, you will typically be warned in a `Caveats` block. You can also see the caveats block by doing `brew info <formula>`.

If you find that there are some tweaks you need to do that isn't documented upstream or explained in a caveats block, then there is a problem with either the upstream package or the Homebrew formula.

Homebrew Python features a caveats block, but that's only there to explain to you how it's been installed to your system. You're free to ignore it if you're fine with using the commands `python3` and `pip3` instead of just `python` and `pip`.

**Comment by memeplex** [comment]
Perhaps I've not expressed it in the most clear way, sorry. What I meant was that maybe say you're building some MacOS app with some standard tooling (I apologise for the vagueness in this description) and this app requires to compile and link against python but, because of following recommendations, it would search for a python framework in /Library/Frameworks, so you would have to tweak its build recipe to point to the atypical /usr/local location. For the case of a *nix build, of course, everything is usually setup to search in /usr/local, so it's fine, my concern is with some hypothetic appleish build. Again, I don't know if this story is realistic, but it's a reasonable inference I'm doing from reading Apple documentation of frameworks.

**Reply by carlocab** [reply]
I see. I wouldn't worry about that. 

Python is very flexible about where you install it; otherwise, using virtual environments would not be possible (cf. Anaconda Python, which is installed at `$HOME/anaconda3`). Virtual environments are in common enough use that standard build recipes should adjust easily to non-standard Python installations. If it doesn't, then it won't just be Homebrew Python users who will have trouble using it.

**Reply by bazhenov** [reply]
You may rest assured, your history is not fictional. It happens on a GitHub Action runners when building applications with embedded Python. GHA runners link such applications over `/Library/Frameworks/Python.framework`. While "brewed" python's framework is in `/usr/local/Frameworks/Python.framework`.

To run such application it required for client to set environment variable `DYLD_FALLBACK_FRAMEWORK_PATH=/usr/local/Frameworks`. Maybe this variable should be initialized automatically when Homebrew is installed?

**Reply by carlocab** [reply]
Setting `DYLD_*_PATH` variables is not a reliable mechanism because it is cleared from the environment by executables protected by SIP:
```
❯ DYLD_FALLBACK_FRAMEWORK_PATH=/usr/local/Frameworks /bin/bash -c 'echo ${DYLD_FALLBACK_FRAMEWORK_PATH-empty}'
empty
❯ FOO=/usr/local/Frameworks /bin/bash -c 'echo ${FOO-empty}'
/usr/local/Frameworks
❯ unset FOO; /bin/bash -c 'echo ${FOO-empty}'
empty
```
If you find need to use `DYLD_*_PATH` variables there's a good chance that your application wasn't linked correctly and that you may actually end up loading a library that is not ABI-compatible.
